tee-worker/core/direct-rpc-client/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+pub mod sgx_reexport_prelude {
+	pub use rustls_sgx as rustls;
+	pub use tungstenite_sgx as tungstenite;
+	pub use url_sgx as url;
+	pub use webpki_sgx as webpki;
+}
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+extern crate alloc;
+
+use alloc::format;
+
+use core::str::FromStr;
+
+use log::debug;
+
+use serde_json::from_str;
+
+use itp_rpc::{Id, RpcRequest, RpcResponse, RpcReturnValue};
+
+use itp_types::RsaRequest;
+use itp_utils::{FromHexPrefixed, ToHexPrefixed};
+
+use litentry_primitives::AesRequest;
+use std::{
+	boxed::Box,
+	error::Error,
+	net::TcpStream,
+	string::{String, ToString},
+	sync::{
+		mpsc::{channel, Sender, SyncSender},
+		Arc,
+	},
+	time::Duration,
+	vec,
+	vec::Vec,
+};
+use tungstenite::{client_tls_with_config, stream::MaybeTlsStream, Connector, Message, WebSocket};
+use url::Url;
+use webpki::{DNSName, DNSNameRef};
+
+pub type Response = (Id, RpcReturnValue);
+
+pub struct IgnoreCertVerifier {}
+
+impl rustls::ServerCertVerifier for IgnoreCertVerifier {
+	fn verify_server_cert(
+		&self,
+		_: &rustls::RootCertStore,
+		_: &[rustls::Certificate],
+		_: DNSNameRef<'_>,
+		_: &[u8],
+	) -> Result<rustls::ServerCertVerified, rustls::TLSError> {
+		log::warn!("Using NoCertVerifier");
+		Ok(rustls::ServerCertVerified::assertion())
+	}
+}
+
+impl rustls::ClientCertVerifier for IgnoreCertVerifier {
+	fn client_auth_root_subjects(
+		&self,
+		_sni: Option<&DNSName>,
+	) -> Option<rustls::DistinguishedNames> {
+		None
+	}
+
+	fn verify_client_cert(
+		&self,
+		_presented_certs: &[rustls::Certificate],
+		_sni: Option<&DNSName>,
+	) -> Result<rustls::ClientCertVerified, rustls::TLSError> {
+		Ok(rustls::ClientCertVerified::assertion())
+	}
+}
+
+pub trait RpcClientFactory {
+	type RpcClient: RpcClient;
+	fn create(
+		&self,
+		url: &str,
+		responses_sink: SyncSender<Response>,
+	) -> Result<Self::RpcClient, Box<dyn Error>>;
+}
+
+pub struct DirectRpcClientFactory {}
+
+impl RpcClientFactory for DirectRpcClientFactory {
+	type RpcClient = DirectRpcClient;
+
+	fn create(
+		&self,
+		url: &str,
+		responses_sink: SyncSender<Response>,
+	) -> Result<Self::RpcClient, Box<dyn Error>> {
+		DirectRpcClient::new(url, responses_sink)
+	}
+}
+
+pub trait RpcClient {
+	fn send(&mut self, request_id: String, params: RequestParams) -> Result<(), Box<dyn Error>>;
+}
+
+pub struct DirectRpcClient {
+	requests_sink: Sender<String>,
+}
+
+impl DirectRpcClient {
+	pub fn new(url: &str, responses_sink: SyncSender<Response>) -> Result<Self, Box<dyn Error>> {
+		let ws_server_url =
+			Url::from_str(url).map_err(|e| format!("Could not connect, reason: {:?}", e))?;
+		let mut config = rustls::ClientConfig::new();
+		// we need to set this cert verifier or client will fail to connect with following error
+		// HandshakeError::Failure(Io(Custom { kind: InvalidData, error: WebPKIError(UnknownIssuer) }))
+		config.dangerous().set_certificate_verifier(Arc::new(IgnoreCertVerifier {}));
```

Comment:
The knee-jerk reaction here would be that we'd rather want to install the relevant cert in the chain here to make the TLS handshake work; however I don't see a way to make it work (esp. on the browser!) without allowing a malicious worker operator to present a cert from _outside_ the enclave :disappointed: 

If that's the case tho, I think it's a tradeoff we'd have to document in big bold flaming letters (both in the code and in relevant docs) :thinking:

Reply:
I think yes but it would require:

* publishing generated cert to enclave registry so it can be added to client config's root store before establishing TLS connection
* adjusting generated cert common name to match the hostname of published trusted rpc endpoint url (currently hardcoded to `Enclave`)

Reply:
Not sure it's worth the trouble tbh.

To prevent an MITM from a malicious node operator, we'd have to make sure that "our" CA cert is _the only one_ in the store when establishing the connection --- otherwise, the operator can present any valid cert for the same domain name, and the TLS logic will be none the wiser.

That _might_ be doable in the (Rust) client, but AFAICT there's no way to do it in the browser (and I'd be more than happy to be proven wrong on that one) --- that means, we still cannot simply rely on TLS there to provide the needed authentication and encryption. If that's the case, I think it just makes sense to stick with the same solution for both.

Reply:
Yeah, our loose concept we had for this was the the enclave's signing key should be part of the certificate, which can be checked against the public registry. This custom approach is of course doable in rust, but browsers obviously don't have such a feature. We have been thinking about a browser extension that would take care of this check though.

But then again how do you explain that to the enduser? To many people don't understand https and the lock XD, we are once again ahead of time.